### PR TITLE
[Draft] Supported platforms

### PR DESCRIPTION
**Pull request by Hoverbear**
I was talking to @nrc the other day about some of our recent  PRs and issues around architecture support etc. He encouraged me to submit this RFC so thanks to him :)

This RFC proposes we formalize some of our platform support and give a direction for any future work on supporting new platforms.

I have an agenda here as I want to support Linux MUSL builds.

### CODE REVIEW

File: text/2019-08-21-supported-platforms.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+This RFC proposes the following terms to be defined for TiKV:
+
+* **Deployment Platforms:** Deployment targets for release binaries produced by
+  the build process. These platforms can *run* TiKV, but not necessarily *build*
+  it. This platform is the concern of end user community.
+* **Development Platforms:** Development targets for interactively developing
+  TiKV on. These platforms can build, debug, and run TiKV, but may not have all
+  features to run a production workload. These platforms are only the concern of
+  our development community.
+* **Tier 1:** Official build processes build, test, and benchmark these targets,
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: I would split the tier descriptions between deploy and dev platforms, since the requirements are somewhat different

File: text/2019-08-21-supported-platforms.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  and they are considered endorsed by the maintainers for all usage.
+      + Example: CentOS 7, our official build OS, on X86-64 with SSE4
+      + Full instructions present and regularly tested
+      + Tested on each PR
+      + Binaries provided
+* **Tier 2:** Official build processes build, but *do not require testing or
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: "Official build processes build" I wouldn't say this, it means we'd be obliged to build for these platforms on CI, which we probably shouldn't do

File: text/2019-08-21-supported-platforms.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      + The target can pass TiKV's test suite and benchmark suite
+      + The proposal is submitted by a regular contributor to TiKV, and they
+        state they intend to maintain support
+* **Tier 2 to Tier 1:**
+      + The target has reliably passed TiKVs test and benchmark suites for a
+        period of 3 months
+      + The target can use an interactive debugger on TiKV and documentation is present
+      + The target supports at least one fuzzer and fuzzing has been performed
+      + The target is used as a development machine by at least one maintainer
+        or three regular contributors
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: These all sound like deployment constraints rather than dev ones

File: text/2019-08-21-supported-platforms.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+### Tier Definitions
+
+This RFC proposes the following definitions for the tiers of platforms and targets:
+
+* **Deployment Platforms**
+      + **Tier 1**:
+        - CentOS 7.x.x on x86-64 on x84-64 with SSE4.2
+        - Ubuntu 16.04 through 19.04 on x84-64 with SSE4.2
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: AIUI, this adding official support, are we comfortable doing it?

File: text/2019-08-21-supported-platforms.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+This list does not attempt to list all posibilities, it lists all major targets
+we are asked about on a regular basis and special cases we are aware of. If
+something is not listed, it should be considered untested and in the "Not
+supported" tier.
+
+### RFC Effects
+
+If accepted, this RFC would:
+
+* **discourage** adding this data to the TiKV documentation.
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: Does "documentation" here mean specifically outside of the repo?

File: text/2019-08-21-supported-platforms.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+This RFC as currently defined does not change any of our *current* support, it
+only clarifies them and defines a path forward for new targets.
+
+## Drawbacks
+
+* We must maintain this compatability specification and ensure correctness
+* Contributors may feel encouraged to try to promote certain platforms up a tier
+  and require us to do more testing on releases.
+* This list does not talk about hardware recommendations like CPUs and memory
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: It does mention x86 and ARM, and we might want to say that bigendian archs will never be supported

File: text/2019-08-21-supported-platforms.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  https://tikv.org/docs/3.0/tasks/deploy/introduction/ for system specification
+  guidelines around these topics.
+
+## Alternatives
+
+* We could merge development and deployment tiers, but this would not be the
+  full story, since TiKV builds in more places than we support deployment.
+* We could not specify what we support and let our users enjoy a fun guessing game
+* We could only specify supported (Tier 1) targets and not worry about partial support.
+* We could limit our Tier 1 to platforms we run Jepsen/Schrodinger tests.
```
Comment (nrc) [inline text/2019-08-21-supported-platforms.md]: Is this not the same as CentOS and Ubuntu?

File: text/2019-08-21-supported-platforms.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
+# Supported Platforms
+
+## Summary
+
+Formally define TiKV's supported build and deployment platform support. Modelled
+after https://forge.rust-lang.org/platform-support.html, we define 3 tiers of
+support. Tier 1 support defines "guaranteed to work" or "well tested" targets,
+tier 2 defines "gauranteed to build" or "ok to experiment with" targets, and
+tier 3 defines "limited support" or "ok with issues" targets.
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: So you'd prefer no difference in support levels between some non-CentOS/Ubuntu Linux and, say, Windows?

File: text/2019-08-21-supported-platforms.md (hunk 4 above)
Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: @c4pt0r do you know?

File: text/2019-08-21-supported-platforms.md (hunk 3 above)
Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: I would definitely not like to see a development platform reach "Officially Supported" status without all of these present.

File: text/2019-08-21-supported-platforms.md (hunk 5 above)
Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: Outside of this RFC.

File: text/2019-08-21-supported-platforms.md (hunk 6 above)
Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: I'll add it to the list of not supported. :)

File: text/2019-08-21-supported-platforms.md (hunk 7 above)
Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: As far as I understand we only run Jepsen on CentOS right now.

File: text/2019-08-21-supported-platforms.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (nrc) [reply text/2019-08-21-supported-platforms.md]: I'd prefer no *formal* difference. I think we can explain to people informally that some might work and some definitely won't, but I think the higher order bit is simply whether one can expect a platform to run TiKV or not.

File: text/2019-08-21-supported-platforms.md (hunk 3 above)
Comment (nrc) [reply text/2019-08-21-supported-platforms.md]: I don't think any of the dev platforms in this RFC (except CentOS) satisfy all these constraints (fuzzing in particular seems a big ask, I would expect us to only fuzz one platform)

Comment (Hoverbear) [reply text/2019-08-21-supported-platforms.md]: Fuzzing works on all Linuxes listed just fine and last I checked it was possible to fuzz on mac as well.

Comment (nrc) [reply text/2019-08-21-supported-platforms.md]: But we certainly can't run all tests and benches on OS X. Has anybody ever tried running OS X and Linux in the same cluster? (and these are all just tier 2 things).

"The target has reliably passed TiKVs test and benchmark suites for a period of 3 months" - given the number of flaky tests, I would argue no platform meets this requirement.

What I meant about fuzzing is that I don't think it was necessary. Hypothetically, if Windows satisfied all the other requirements, but couldn't fuzz, I'd be happy for it to be a supported dev platform because we can fuzz elsewhere.

In other words, I don't think you need to run TiKV to develop it, you just need to be able to build for a platform where you can run it (by analogy, in mobile development, there are deployment requirements for phones, but you are not expected to develop on one, and the machine you use to develop on does not need to run phone apps).

**Comment by Hoverbear** [comment]
Closing. I am not longer involved with the project and have no stake in this merging. Feel free to create a new PR.
